### ISSUE TITLE
Help needed with pass-through.

### ISSUE THREAD

**Issue by 2WhlWzrd** [koying/SPMC#947]
I know that this is not the preferred setup, but it is what I have and it worked with earlier versions of Kodi 17, so I'm sure there is a way to make it work with the more versatile SPMC 17.
My AVR does not have an HDMI input, it does however have S/PDIF Coax input and my TV has an S/PDIF Coax output. So I plug my FireTV stick into an HDMI port on the TV, the TV has 2 settings for the digital output, "Raw" or "PCM", I use Raw. Then connect the Digital Audio output of the TV to the Digital Audio input of the AVR.
Using early versions of Kodi 17 with default audio settings (pass-through enabled), I would get 5.1 audio from AC3 encoded files, but had to enable transcoding to get 5.1 audio from either E-AC3 or DTS
encoded tracks. 

[kodi 17 beta2 log](https://pastebin.com/4Vb1uBnr)
Refresh rate switching was broken in this version.

It is my understanding that Scott Davilla coded the AE for Krypton and the audio worked perfectly upto version [kodi-20160901-ae5fc7c-master-armeabi-v7a.apk](http://ftp.igh.cnrs.fr/pub/xbmc/nightlies/android/arm/kodi-20160901-ae5fc7c-master-armeabi-v7a.apk) after which any device compatability coding was removed in favor of Android standards only coding.
I don't think this was the standard PCM hack because I have tried this with Kodi and SPMC versions before 17 and even the Kodi 17 "shitty" test build. None of them work.
So whatever made it work was removed after the version listed above. 

Using SPMC 17 with [recommended audio settings](https://github.com/koying/SPMC/wiki/Recommended-passthrough-audio-settings-for-SPMC%2C-Fire-TV-and-AV-Receivers) from the SPMC Wiki.
I only get 2.0 audio regardless of encoding.

[spmc 17.4 rc1 log](http://paste.semperpax.com/p5vg9qekl)

Are there alternative settings for the current SPMC 17 that will make passthrough work with my setup?
or can we figure out what made it work with the early versions of Kodi 17 and include it in SPMC?
I would really like to have a more current version with working refresh rate switching and audio pass-through.

**Comment by MrMC** [comment]
not really, AE has burned through several authors. I (davilla) just did the original AudioTrack sink and adapted the raw passthrough that koying for SPMC 16 to the current AudioTrack sink from Krypton for MrMC.

**Comment by classicjazz** [comment]
The Wiki article was written assuming an HDMI connection to the AVR. In your case, you have a FireTV stick, which is very limited in what it can pass through in the first place. And you have SPDIF, which is limited to either 2 channel PCM or lossy surround sound, regardless of what the Stick can do. 

So, I would just use SPMC's software decoding to AC3 surround sound for all audio formats.

**Comment by 2WhlWzrd** [comment]
@MrMC 
It's been a while back, but I think the post that I got the information from regarded changes to the Kodi AE in 17, and maybe more specifically the AudioTrack sink  which is why you where the only author mentioned.
So apparently I have misinterpreted the information, thanks for the clarification.
But if you could help shed some light on why the above version works the way it does, please do.

@classicjazz 
If I'm understanding "SPMC's software decoding to AC3 surround sound for all audio formats" correctly,
that is exactly what the version of Kodi listed above is doing. But what audio settings do I use to achieve this in SPMC?
The settings I am using in Kodi version above are:
   Audio output device = Android, audiotrack [not changeable in this version]
   Number of channels = 2.0
   Output configuration = Best Match
   Stereo upmix = off
   Maintain original volume on downmix = on
   Boost centre channel when downmixing = 0 dB
   Resample quality = High
   Keep audio devices alive = 10 Minutes [this will prevent your AVR from going to sleep]
   Enable audio DSP processing = off
   Enable passthrough = on
   Passthrough output device = Android, audiotrack [not changeable in this version]
   Dolby Digital (AC3) capable receiver = on
   Enable Dolby Digital (AC3) transcoding = on
   Dolby Digital Plus (E-AC) capable receiver = [n/a in this version]
   DTS capable receiver = off 

With these settings in SPMC, aside from Audio and Passthrough output devices, all I get is silence.
I would upload a debug log, but I keep getting "Could not read log" error after playing a test file.

**Comment by MrMC** [comment]
MrMC is a hybrid of the 16/17 codebase plus our code additions for native emby/plex clients and other fun things. The player and AE code base is 17 up to a certain point where we feel its behavior and performance became compromised by the further changes.

MrMC works for both IEC and RAW passthrough because we took the concept of using RAW (from SPMC) and applied it in a different way to core AE/17. We felt that while having multiple devices (normal, IEC and RAW) was clever, it complicated user settings too much and we figured out a way do this with single IEC/RAW setting. The AudioTrack sink will figure out the rest.

Kodi devs could certainly backport our changes but they stubbornly refuse to do it, claiming that is can cause damage equipment and hurt your ears. This is just plain silly. Doing RAW passthrough has been around for decades and was (at one time), the only way to do passthrough. Sure, if you have the volume turned way up and/or your code base is faulty and does not handle the internal volume properly, you can get a blast of noise. That is why when testing or setting up on a new device, you lower the volume until you figure out if you need IEC or RAW :)

MrMC actively tests on iOS/tvOS, Amazon FireTV1/2, FireSticks, FireTablets and the Nvidia Shield TV. Pretty sure Koying tests on the same (except of iOS/tvOS). Kodi devs only seem to care about  the Nvidia Shield and would rather point fingers at other devices and claim faulty firmware. 

MrMC takes a different approach, following standard APIs is important but sometime as coders, one must stray outside the 'standard' and code 'quirk' handlers to provide a good user experience. Just a fact of life in coding for real world devices, nothing is 100 percent perfect. Deal with it.

**Comment by BigOkie** [comment]
Not a big fan however of charging for the app.  Neither SPMC nor Kodi does this.  I pass.

**Comment by MrMC** [comment]
I think the discussion was about MrMC vs SPMC vs Kodi and passthrough differences.  Nothing like a drive by thread highjack :)

Kodi does not and will never again passthrough using RAW, if this floats your boat, fine. If not, then SPMC is your man.

**Comment by classicjazz** [comment]
I don't know the specifics about your TV. But I assume that RAW means your TV passes through the bitstream and PCM means that it decodes the bitstream to PCM. So, I would use RAW for your TV SPDIF setting, with the expectation that it will passthrough AC3 to the AVR that way.

Back to SPMC:
* Select either 7.1 or 5.1 audio channels, as appropriate for your AVR & speaker setup
* Enable passthrough = off (disabled)
* Passthrough output device = Android, audiotrack [not changeable in this version]
* Dolby Digital (AC3) capable receiver = on
* Enable Dolby Digital (AC3) transcoding = on
* Dolby Digital Plus (E-AC) capable receiver = [n/a in this version]
* DTS capable receiver = off

Fingers crossed, you should now be transcoding everything to 5.1/7.1 surround sound in AC3 format. Unfortunately, I don't own a FireTV or have your TV/AVR to independently test. But it should be similar to using a 3.1/5.1 soundbar that doesn't support HDMI ARC.

**Comment by 2WhlWzrd** [comment]
@classicjazz 
Your assumptions about the TV are correct. From earlier testing that I have done, the Raw setting will allow multichannel, while the PCM setting only allows 2 channel audio.
As far as your recommended settings go;
When changing "Number of channels" to anything above 2.0, the option to "Enable Dolby Digital (AC3) transcoding" is removed.
When setting "Enable passthrough" to off, the setting for "Dolby Digital (AC3) capable receiver" and all other sub-settings are disabled (greyed out).
Leading me to believe that these settings are having no effect on the output.
But the results were only 2 channel audio regardless.

My settings in this version of Kodi, I believe are allowing AC3 passthrough and transcoding everything else.
Looking at the [Audio quickstart guide](http://kodi.wiki/view/Audio_quickstart_guide), I would expect this to be the correct behavior for these settings.
Maybe this was a bug, that was fixed early in 17, then the patch removed.
So the question still is how do we do this in SPMC?

**Comment by docgonzo2015** [comment]
thats a good question!

**Comment by ians325** [comment]
I would like to the passthrough working my Hisilicon correctly what, info you require as its raw, it does dolby on its bespoke videoplayer. I would like get this issue sorted as millions of hisilcon chip uses with kodi facing the issue with. Im will to run all the tests that you need for the correct information to code this if you help me telling me what info you require thanks.
